cpp/src_prims/sparse/csr.h

```diff
@@ -47,170 +47,204 @@ static const float MIN_FLOAT = std::numeric_limits<float>::min();
+    */
+  void allocate(int nnz, bool init, cudaStream_t stream) {
+    this->allocate(nnz, -1, init, stream);
+  }
 
   /**
-     * @brief Allocate underlying arrays and the size of the square dense matrix
-     * @param nnz: sets the size of the underlying arrays
-     * @param size: number of rows and cols in the square dense matrix
-     * @param init: should arrays be initialized to zeros?
-     */
-  void allocate(int nnz, int size, bool init = true) {
-    this->allocate(nnz, size, size, init);
+    * @brief Allocate the underlying arrays
+    * @param nnz: size of the underlying row/col/val arrays
+    * @param size: the number of rows/cols in a square dense matrix
+    * @param init: should values be initialized to 0?
+    * @param stream: CUDA stream to use
+    */
+  void allocate(int nnz, int size, bool init, cudaStream_t stream) {
+    this->allocate(nnz, size, size, init, stream);
   }
 
   /**
-     * @brief Allocate underlying arrays and the size of the non-square dense matrix
-     * @param nnz: sets the size of the underlying arrays
-     * @param n_rows: number of rows in the dense matrix
-     * @param n_cols: number of cols in the dense matrix
-     * @param init: should arrays be initialized to zeros?
-     */
-  void allocate(int nnz, int n_rows, int n_cols, bool init = true) {
+    * @brief Allocate the underlying arrays
+    * @param nnz: size of the underlying row/col/val arrays
+    * @param n_rows: number of rows in the dense matrix
+    * @param n_cols: number of columns in the dense matrix
+    * @param init: should values be initialized to 0?
+    * @param stream: stream to use for init
+    */
+  void allocate(int nnz, int n_rows, int n_cols, bool init,
+                cudaStream_t stream) {
     this->n_rows = n_rows;
     this->n_cols = n_cols;
     this->nnz = nnz;
-    MLCommon::allocate(this->row_ind, this->nnz, init);
-    MLCommon::allocate(this->row_ind_ptr, this->nnz, init);
-    MLCommon::allocate(this->vals, this->nnz, init);
-  }
-
-  /**
-     * @brief Frees the memory from the underlying arrays
-     */
-  void free() {
-    try {
-      if (row_ind != nullptr) CUDA_CHECK(cudaFree(row_ind));
-
-      if (row_ind_ptr != nullptr) CUDA_CHECK(cudaFree(row_ind_ptr));
-
-      if (vals != nullptr) CUDA_CHECK(cudaFree(vals));
 
-      row_ind = nullptr;
-      row_ind_ptr = nullptr;
-      vals = nullptr;
+    this->rows.resize(this->nnz, stream);
+    this->cols.resize(this->nnz, stream);
+    this->vals.resize(this->nnz, stream);
 
-    } catch (Exception &e) {
-      std::cout << "An exception occurred freeing COO memory: " << e.what()
-                << std::endl;
+    if (init) {
+      cudaMemsetAsync(this->row_ind.data(), 0, this->nnz * sizeof(Index_Type),
+                      stream);
+      cudaMemsetAsync(this->row_ind_ptr.data(), 0,
+                      this->nnz * sizeof(Index_Type), stream);
+      cudaMemsetAsync(this->vals.data(), 0, this->nnz * sizeof(T), stream);
```

Comment:
`CUDA_CHECK()` missing.

Reply:
In fact, how about creating a private method to initialize all these to zero? Since similar such initialization also happens above in the ctor.